CrateDB output plugin

felixge opened the pull request:
Hello,

this PR implements an output plugin for [CrateDB](http://crate.io/) using their [Postgres Wire Protocol](https://crate.io/docs/crate/reference/protocols/postgres.html).

Please let me know what kind of changes, if any, you'd like me to make in order to get this merged.

Many thanks in advance 😊
Felix

### Required for all PRs:

- [x] Signed [CLA](https://influxdata.com/community/cla/).
- [x] Associated README.md updated.
- [x] Has appropriate unit tests.

danielnelson commented:
Thanks for the pull request @felixge, I'm pretty far behind on reviews, so it might take me a little while before I am able to review.  One thing I notice though is that the plugin uses a different postgres library than the one we use in [existing plugins](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/postgresql/postgresql.go#L12), is it possible to switch to pgx?

felixge commented:
@danielnelson good idea, I didn't even notice pgx was used already, as I was just looking at other output plugins.

That being said, unfortunately using pgx causes my tests to fail with the following error:

```
--- FAIL: TestConnectAndWrite (0.02s)
        Error Trace:    cratedb_test.go:30
	Error:      	Received unexpected error:
	            	ERROR: TableUnknownException: Table 'test.pg_type' unknown (SQLSTATE XX000)
```

This is because pgx [queries the postgres catalog](https://github.com/jackc/pgx/blob/master/conn.go#L387-L393) when connecting, and CrateDB doesn't implement those catalog tables. As far as I can tell this feature cannot be disabled in pgx.

So if possible, I'd like to keep using `lib/pq` for this plugin. Let me know what you think.

Cheers
Felix

danielnelson commented:
Thanks for opening the upstream issue.  Lets try to get it resolved in pgx so we can avoid increasing the size of the Telegraf binary.  In the meantime it makes sense to leave your pull request with pq in case anyone would like to do their own testing.

felixge commented:
@danielnelson I've worked with upstream to add CrateDB support to [pgx](https://github.com/jackc/pgx/pull/336). I also upgraded this PR to use pgx instead of lib/pq. This caused a few postgres tests to fail due to changes in the [way newer version of pgx do type casting](https://github.com/jackc/pgx/issues/343).

Please let me know how you feel about the PR at this point. Are the changes to the postgres test acceptable? If yes I'm happy to do some final cleanups (some comments in the code still talk about lib/pq) to get this merged.

danielnelson commented:
The tests should be fine, we convert all integers to int64, though I'm interested to hear if the change was on purpose or not.

Glancing over the code, there is a lot of manual escaping.  Can we reduce this by using [NamedArg](https://golang.org/pkg/database/sql/#NamedArg)?  Is the object type specified as JSON, maybe we can use the json module?

felixge commented:
> Glancing over the code, there is a lot of manual escaping. Can we reduce this by using NamedArg?

Unfortunately not. CrateDB currently doesn't support the full postgres wire protocol, which unfortunately includes placeholder variables.

> Is the object type specified as JSON, maybe we can use the json module?

No it looks similar, but it's not JSON. See https://crate.io/docs/crate/reference/en/latest/sql/data_types.html#object-literals

Please let me know how to proceed.

danielnelson commented:
I wonder if this is related to the implicit prepared queries made by pgx? https://github.com/jackc/pgx/issues/331  I always thought that the named args were expanded on the client side, but if a prepared statement is being used then that would not be the case.  If this is true, perhaps we can hook into the code to do expansion. Maybe like this? https://github.com/jackc/pgx/blob/master/query.go#L535

I'm not sure what to do about the object literals other than have extra care testing that the escaping is correct.

felixge commented:
@danielnelson yes, this is related to prepared statements. Prepared statements should always be used when trying to safely inject data into a SQL query. Unfortunately CrateDB doesn't support a certain variant of the [Extended Query format](https://www.postgresql.org/docs/9.6/static/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY):

> Note: A parameter data type can be left unspecified by setting it to zero, or by making the array of parameter type OIDs shorter than the number of parameter symbols ($n) used in the query string.

CrateDB however doesn't determine the number of parameters by looking at the number of placeholders, but instead [looks at the length of the array of parameter type OIDs](https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/protocols/postgres/PostgresWireProtocol.java#L425-L451). Pgx (and lib/pq) are not setting this array (as it's optional), causing CrateDB to always complain about too many parameters being given.

Of course one can always try to inject the data by manually escaping it on the client site (which is what `sanitize.SanitizeSQL` and also my code in this PR) does, but this is always prone to error. But in case of CrateDB, there is no choice.

So the question for you is: Would you prefer to use the escaping/sanitize mechanism form pgx over the code in this PR? If yes, I'll try to change my code accordingly. But it will most likely require giving up on using the `database/sql` interface, because the `sanitize` pkg of pgx is internal, so I can only use it indirectly by using the pgx native APIs ...

Please let me know how you'd like to proceed.

I agree about the object literals. I've added good test coverage for that code to make sure it works.

danielnelson commented:
I guess using sanitize is not an option due to being internal, maybe in the future after pgx gains the ability to not use prepared statements we can update it.

felixge commented:
@danielegozzi yeah. I also think CrateDB will fix the problem preventing prepared statement support on their end at some point. That will be even better.

Anyway, do you think we can move forward with this PR? If yes, are there any remaining improvements you'd like me to make?

danielnelson commented:
I'll try to do a closer review soon, shouldn't be a problem to finish this for 1.5.

danielnelson commented on plugins/outputs/cratedb/cratedb.go:
```diff
@@ -0,0 +1,202 @@
+		// -> ERROR:  SQLParseException: For input string: "14305102049502225714"
+
+		cols := []interface{}{
+			int64(m.HashID()),
```
I hadn't considered the values of this hash to be part stable across versions, would it be a problem if this hash changed even if the series key (measurement name + tagset) did not?

felixge replied:
Np. I think a cryptographic hash function makes more sense for the use case (deduplication) anyway. Let me know if you think https://github.com/influxdata/telegraf/pull/3210/commits/626f6532dbd2b5f7395e49d69d2ed8b9ee02b30d works for you.

danielnelson commented on plugins/outputs/cratedb/cratedb.go:
```diff
@@ -0,0 +1,202 @@
+
+		cols := []interface{}{
+			int64(m.HashID()),
+			m.Time().In(loc),
```
Why in local time, what about just sending all times in UTC?

felixge replied:
CrateDB converts to UTC anyway, but you're right, this actually makes things a bit simpler. https://github.com/influxdata/telegraf/pull/3210/commits/e2460329bb4820af9e9d009e5e4495fb392316b7

danielnelson commented on plugins/outputs/cratedb/cratedb_test.go:
```diff
@@ -0,0 +1,138 @@
+		{map[string]interface{}{}, `{}`},
+		{map[string]interface{}(nil), `{}`},
+		{map[string]interface{}{"foo": "bar"}, `{"foo" = 'bar'}`},
+		{map[string]interface{}{"foo": "bar", "one": "more"}, `{"foo" = 'bar', "one" = 'more'}`},
```
Can you add some extra test cases for string escaping, here are some ideas:
```go
map[string]interface{}{"fo"o": "b'ar", "ab'c": `xy"z`, `on"""e`: "mo'''re"}
```

felixge replied:
I've added the additional test case https://github.com/influxdata/telegraf/pull/3210/commits/c43432203e0fe64d2b274c43d02c58d0bfe248c6 and a new smoke test that validates that the escaped value is a valid SQL literal according to CrateDB: https://github.com/influxdata/telegraf/pull/3210/commits/8b85f0bbbe7eb4cfbcc26e22d94a639eb7b15a8b

danielnelson commented on Makefile:
```diff
@@ -86,6 +86,12 @@ docker-run:
 		-e SLAPD_CONFIG_ROOTPW="secret" \
 		-p "389:389" -p "636:636" \
 		-d cobaugh/openldap-alpine
+	docker run \--name cratedb \
```
Do we need to remove the backslash here? `\--name`

felixge replied:
3b22f2eefaf9553dbc653fce135893f37d213707

danielnelson commented on plugins/outputs/cratedb/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Table Schema
+
+The plugin requires a a table with the following schema.
```
Extra `a` in this sentence

felixge replied:
https://github.com/influxdata/telegraf/pull/3210/commits/6c2df9e9c765cea1a8631e12dcd828b40f58b467

danielnelson commented on plugins/outputs/cratedb/docker-compose.yaml:
```diff
@@ -0,0 +1,7 @@
+version: "3"
```
What do we need the docker compose file for?

felixge replied:
I was using it for local testing. Shouldn't have checked it in. It's gone as of https://github.com/influxdata/telegraf/pull/3210/commits/a54054f906f0835731e6757b277e2e54852fe788

felixge commented:
@danielnelson Thanks for taking the time to review this PR. I've addressed all concerns you raised, please let me know if you'd like me to make any additional changes.

For some reason some circleci tests failed:

```
--- FAIL: TestWriteHTTPSNoClientAuth (0.01s)
	Error Trace:	http_listener_test.go:233
	Error:      	Received unexpected error:
	            	Post https://localhost:33545/write?db=mydb: x509: certificate has expired or is not yet valid
2017/11/02 19:58:12 I! Started HTTP listener service on :0
2017/11/02 19:58:12 http: TLS handshake error from 127.0.0.1:36421: remote error: tls: bad certificate
2017/11/02 19:58:12 I! Stopped HTTP listener service on  :0
--- FAIL: TestWriteHTTPSWithClientAuth (0.02s)
	Error Trace:	http_listener_test.go:247
	Error:      	Received unexpected error:
	            	Post https://localhost:42092/write?db=mydb: x509: certificate has expired or is not yet valid
...
FAIL
FAIL	github.com/influxdata/telegraf/plugins/inputs/http_listener	6.757s
```

But I think those are unrelated to my changes.

felixge commented:
I just merged master into this branch, and there is now an additional test failure:

```
--- FAIL: TestRateLimiter (0.25s)
	<autogenerated>:1: 
                          
	Error Trace:	limiter_test.go:26
		
	Error:      	Not equal: 
		
	            	expected: 5
		
	            	received: 0
FAIL
FAIL	github.com/influxdata/telegraf/internal/limiter	0.523s
```

That being said, the latest commit on master is also failing CircleCI right now, so I still think those failures are unrelated to this PR.

danielnelson commented:
I fixed the http_listener issue and it should pass now if you rebase.

felixge commented:
@danielegozzi I merged the latest changes from master into this PR and got b0d6723cdaf4a2f77b0fb3964ed54dc1d161b246 to pass the circleci checks.

Please let me know if this makes this PR mergeable or if you'd like to see some more improvements.

felixge commented:
@danielnelson thank you so much for taking the time to review and merge this PR. 🙏
